jib-core/src/test/java/com/google/cloud/tools/jib/image/LayerBuilderTest.java

```diff
@@ -96,19 +119,54 @@ public void testBuild() throws URISyntaxException, IOException {
                   throw new RuntimeException(ex);
                 }
               });
+    }
+  }
 
-      // Verifies that blobA was added.
-      TarArchiveEntry header = tarArchiveInputStream.getNextTarEntry();
-      String expectedExtractionPath = "extract/here/blobA";
-      Assert.assertEquals(expectedExtractionPath, header.getName());
-
-      String expectedFileString = new String(Files.readAllBytes(blobA), StandardCharsets.UTF_8);
-
-      String extractedFileString =
-          CharStreams.toString(
-              new InputStreamReader(tarArchiveInputStream, StandardCharsets.UTF_8));
+  @Test
+  public void testToBlob_reproducibility()
+      throws IOException, NoSuchAlgorithmException, InterruptedException {
+    Path testRoot = temporaryFolder.getRoot().toPath();
+    Path root1 = Files.createDirectories(testRoot.resolve("files1"));
+    Path root2 = Files.createDirectories(testRoot.resolve("files2"));
+
+    String extractionPath = "/somewhere";
+
+    String contentA = "abcabc";
+    Path fileA1 = createFile(root1, "fileA", contentA, 10000);
+    Path fileA2 = createFile(root2, "fileA", contentA, 20000);
+    String contentB = "yumyum";
+    Path fileB1 = createFile(root1, "fileB", contentB, 10000);
+    Path fileB2 = createFile(root2, "fileB", contentB, 20000);
+
+    // check if modified times are off
+    Assert.assertNotEquals(Files.getLastModifiedTime(fileA1), Files.getLastModifiedTime(fileA2));
+    Assert.assertNotEquals(Files.getLastModifiedTime(fileB1), Files.getLastModifiedTime(fileB2));
+
+    // create layers of exact same content but ordered differently and with different timestamps
+    Blob layer =
+        new LayerBuilder(Arrays.asList(fileA1, fileB1), extractionPath, true).build().getBlob();
+    Blob reproduced =
+        new LayerBuilder(Arrays.asList(fileB2, fileA2), extractionPath, true).build().getBlob();
+
+    // this is a control layer that is the same as 'layerReproduced' without reproducibility on
+    Blob notReproduced =
+        new LayerBuilder(Arrays.asList(fileB2, fileA2), extractionPath, false).build().getBlob();
+
+    MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
+
+    byte[] layerHash = sha256.digest(Blobs.writeToString(layer).getBytes());
+    byte[] reproducedHash = sha256.digest(Blobs.writeToString(reproduced).getBytes());
+    byte[] notReproducedHash = sha256.digest(Blobs.writeToString(notReproduced).getBytes());
+
+    Assert.assertThat(layerHash, CoreMatchers.is(reproducedHash));
+    Assert.assertThat(layerHash, CoreMatchers.not(notReproducedHash));
+  }
 
-      Assert.assertEquals(expectedFileString, extractedFileString);
-    }
+  private Path createFile(Path root, String filename, String content, long lastModifiedTime)
+      throws IOException {
+    // TODO : add file permissions and other stuff, to make sure we're stripping them.
```

Comment:
File permissions probably shouldn't be stripped right? They probably shouldn't vary between builds on different machine.

Reply:
Oops, I meant ownership not permissions.

Reply:
Ah okay, what should be the stripped version of ownership? Just like root?